pkgs/applications/science/math/mathematica/default.nix

```diff
@@ -103,6 +119,9 @@ in stdenv.mkDerivation {
     "--set USE_WOLFRAM_LD_LIBRARY_PATH 1"
     # Fix xkeyboard config path for Qt
     "--set QT_XKB_CONFIG_ROOT ${xkeyboard_config}/share/X11/xkb"
+  ] ++ lib.optionals cudaSupport [
+    "--set CUDA_PATH ${cudaEnv}"
+    "--set NVIDIA_DRIVER_LIBRARY_PATH ${cudaEnv}/lib/libnvidia-tls.so"
```

Comment:
Is there a reason not to just get this from `/run/opengl-driver/`? Creating a dependency on nvidia_x11 means that users can't bring their own NVIDIA driver. This could cause hiccups when the stable driver doesn't work on their system, either bc it's too old (V100/A100 models) or too new (legacy cards).

I'm also not sure how this will work on non-NixOS machines.

Reply:
>Is there a reason not to just get this from `/run/opengl-driver/`?

I'd imagine that should work fine on NixOS, yes.

> Creating a dependency on nvidia_x11 means that users can't bring their own NVIDIA driver. This could cause hiccups when the stable driver doesn't work on their system, either bc it's too old (V100/A100 models) or too new (legacy cards).

Unless I'm mistaken, this will accept any of the nvidia_x11* packages, not just the stable version.

>I'm also not sure how this will work on non-NixOS machines.

I'm hoping that If the provided nvidia_x11 package version matches the installed driver, everything will work out, but I don't currently have a machine to test this theory with.

Reply:
It's still better to use `addOpenGLRunpath` instead, unless for some reason it doesn't work:

- It's conventional
- You won't have to rebuild the package for every different piece of hardware
- I second @samuela, I'm not sure if it would be possible to override the driver via `LD_LIBRARY_PATH` on non-NixOS if you'd already have matching libraries in the rpath

Reply:
> I'm hoping that If the provided nvidia_x11 package version matches the installed driver, everything will work out, but I don't currently have a machine to test this theory with.

According to my tests, the `nvidia_x11` package must be supplied with a matching version to the installed one. I tried different versions, but all of them didn't work out.

Reply:
Yeah this is why `addOpenGLRunpath` is the recommended way of linking with the Nvidia driver. User-mode libraries like `cudatoolkit`, `cudnn`, `cutensor`, etc we link in normally.

Reply:
`addOpenGLRunpath` can't remove nvidia_x11 as a dependency for `cudaEnv`/`CUDA_PATH`, so it's not of much use here.

Reply:
`addOpenGLRunpath` simply adds /run/opengl-driver/ to the RUNPATH of binaries. If for some reason you require /run/opengl-driver/ in `CUDA_PATH`, you can do so manually. The `nvidia_x11` driver should never be used as an input to user-level derivations.

Reply:
But note that CUDA != Nvidia driver, so I would be surprised if Mathematica actually looks for driver libraries in `CUDA_PATH`.

Reply:
I see. I don't know anything about how mathematica is packaged and deployed, so maybe you'll help. Are you sure mathematica needs the graphics driver to be in `CUDA_PATH`? What happens if you leave only the cuda libraries in `CUDA_PATH`, but either

- manually `patchelf` all mathematica's executables and libraries to prefix `RPATH` with `/run/opengl-driver/lib`? Something like https://github.com/NixOS/nixpkgs/blob/634141959076a8ab69ca2cca0f266852256d79ee/pkgs/development/libraries/science/math/nccl/default.nix#L34 (arbitrary chunk of code I found with github search)
- or prepend `/run/opengl-driver/lib` in `--prefix LD_LIBRARY_PATH` in `wrapProgramFlags`?

Reply:
I should mention first and foremost that CUDALink (to my knowledge) doesn't utilize the dynamic linker to find CUDA or the Nvidia driver, so rpaths, `LD_LIBRARY_PATH` etc. won't have any effect. Rather, it searches common locations (e.g. /opt for CUDA, /usr/lib for Nvidia driver) and lastly falls back to `CUDA_PATH` and `NVIDIA_DRIVER_LIBRARY_PATH` (which are not colon-separated lists like e.g. `PATH`). It does *also* look for the Nvidia driver under `CUDA_PATH` which is why I bundled them all together in this PR.
I'm not having much luck leaving only CUDA libs in `CUDA_PATH` and setting `NVIDIA_DRIVER_LIBRARY_PATH` elsewhere at the moment, so I will have to spend some time probing CUDALink internals again to figure out the best means of separating the two.

Reply:
Ah, that's v annoying that they don't just follow the `dlopen` standards. A few random things that come to my mind:
* It may be useful to try running with `LD_DEBUG=libs`. Not sure how helpful that will be given that they're looking with non-standard machinery, but worth a shot. If that fails, `LD_DEBUG=all` may be useful. If all else fails, `strace` is the way to go.
* There's nothing wrong with `wrapProgram` per se, just that we can't use `nvidia_x11`. Manually setting `/run/opengl-driver/lib` is totally fair game. It's unfortunate that `addOpenGLRunpath` doesn't work automagically, but that's not the end of the world.
 
Based on your explanation of the situation, I would try something like
```nix
  cudaEnv = buildEnv {
     name = "mathematica-cuda";
     paths = [ cudatoolkit cudatoolkit.lib ];
     pathsToLink = [ "/bin" "/include" "/lib" ];
     postBuild = ''
       ln -s $out/lib $out/lib64
     '';
   };

...

    "--set CUDA_PATH ${cudaEnv}"
    "--set NVIDIA_DRIVER_LIBRARY_PATH /run/opengl-driver/lib/libnvidia-tls.so"
```

Reply:
If _all_ else fails, you might try
```nix
  cudaEnv = buildEnv {
     name = "mathematica-cuda";
     paths = [ cudatoolkit cudatoolkit.lib "/run/opengl-driver/" ];
     pathsToLink = [ "/bin" "/include" "/lib" ];
     postBuild = ''
       ln -s $out/lib $out/lib64
     '';
   };
```
or something along those lines.

Reply:
Also it would be good to open an issue upstream asking how to set custom CUDA and driver locations. They ought to either have a way to accomplish this or acknowledge it as a bug.

Reply:
A [document](https://reference.wolfram.com/language/CUDALink/tutorial/Setup.html#10897850) exists about setting up a custom-installed CUDA; however, it did not work as expected by not creating a combination like `cudaEnv` and setting those variables separately in my testing too.

Reply:
From that document, I'm seeing another variable `CUDA_LIBRARY_PATH`. Did you try setting that variable as well? It appears to be for the location of `libcuda.so` which confusingly is a part of the nvidia driver (available in `/run/opengl-driver`) and not `cudatoolkit`...

Reply:
I was testing the patch, so my attempt did not involve using drivers from `/run/opengl-driver`, but instead used `nvidia_x11`. That combination failed at `InstallCUDA[]`.

Reply:
> AFAIU setting CUDA_PATH and NVIDIA_DRIVER_LIBRARY_PATH separately is how the first version of this PR was written however. Did that version not work as expected?

I found out that it wasn't doing anything so I removed it; it only works since the Nvidia driver is inside `CUDA_PATH` anyways.

Reply:
oh weird... sounds like it's quite buggy, or at least the docs are not correct